docs/petr4spec/Petr4-spec.mdk

```diff
@@ -142,509 +149,2979 @@ p4grammar {
+|   | $\vert$ | $\action$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\entry \pmatch \action$ | $\qquad \text{entries}$ |
+|   | $\vert$ | $\custom \bool \var \exp$ | $\qquad \text{custom property}$ |
+
+**Methods:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\extMethod$ | $::=$ | $\constructor \var \prmDef$ | $\qquad \text{constructors}$ |
+|   | $\vert$ | $\abstractMethod \baseTyp \var \typVar \prmDef$ | $\qquad \text{abstract methods}$ |
+|   | $\vert$ | $\methodDef \baseTyp \var \typVar \prmDef$ | $\qquad \text{methods}$ |
+
+**Declarations:**
+
+|     |      |     |     |
+|~~~~:|:~~~~:|:~~~~|:~~~~|
+| $\dcl$ | $::=$ | $\constDcl \baseTyp \var \exp$ | $\qquad \text{constants}$ |
+|   | $\vert$ | $\inst \baseTyp \args \var \stmts$ | $\qquad \text{instantiations (initialized)}$ |
+|   | $\vert$ | $\instNoBlock \baseTyp \args \var$ | $\qquad \text{instantiations (uninitialized)}$ |
+|   | $\vert$ | $\parserDcl \var \typVars {\overline {\prmDef_1}} {\overline {\prmDef_2}} \dcls \states$ | $\qquad \text{parsers}$ |
+|   | $\vert$ | $\controlDcl \var \typVars {\overline {\prmDef_1}} {\overline {\prmDef_2}} \dcls \stmts$ | $\qquad \text{controls}$ |
+|   | $\vert$ | $\funcDcl \baseTyp \var \typVars \prmDefs \stmts$ | $\qquad \text{functions}$ |
+|   | $\vert$ | $\externFuncDcl \baseTyp \var \typVars \prmDefs$ | $\qquad \text{extern functions}$ |
+|   | $\vert$ | $\varDclInit \baseTyp \var \exp$ | $\qquad \text{local variables (initialized)}$ |
+|   | $\vert$ | $\varDcl \baseTyp \var$ | $\qquad \text{local variables (uninitialized)}$ |
+|   | $\vert$ | $\valSetDcl \baseTyp \exp \var$ | $\qquad \text{parser value sets}$ |
+|   | $\vert$ | $\actionDcl \var \prmDefs \stmts$ | $\qquad \text{actions}$ |
+|   | $\vert$ | $\tableDcl \var \props$ | $\qquad \text{tables}$ |
+|   | $\vert$ | $\headerDcl \typVar \fieldTyps$ | $\qquad \text{header types}$ |
+|   | $\vert$ | $\headerUnionDcl \typVar \fieldTyps$ | $\qquad \text{header union types}$ |
+|   | $\vert$ | $\structDcl \typVar \fieldTyps$ | $\qquad \text{struct types}$ |
+|   | $\vert$ | $\errDcl \fields$ | $\qquad \text{error types}$ |
+|   | $\vert$ | $\matchkindDcl \fields$ | $\qquad \text{match kind types}$ |
+|   | $\vert$ | $\enumDcl \typVar \fields$ | $\qquad \text{enumerated types}$ |
+|   | $\vert$ | $\serEnumDcl \baseTyp \typVar \field \exp$ | $\qquad \text{serializable enumerated types}$ |
+|   | $\vert$ | $\externObjDcl \var \typVars \extMethods$ | $\qquad \text{extern objects}$ |
+|   | $\vert$ | $\typdefDcl \typVar \baseTyp$ | $\qquad \text{type definitions}$ |
+|   | $\vert$ | $\typdefDcl \typVar \dcl$ | $\qquad \text{type definitions (declarations)}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \baseTyp$ | $\qquad \text{generative type definitions}$ |
+|   | $\vert$ | $\newtypeDcl \typVar \dcl$ | $\qquad \text{generative type definitions (declarations)}$ |
+|   | $\vert$ | $\controlTypDcl \typVar \typVars \prmDefs$ | $\qquad \text{control types}$ |
```

Comment:
Not a big fan of the abbreviations here and below, especially as we wrote them out above. Let's be consistent?

Reply:
applied. I wanted to differentiate between an object declaration and a type declaration but we can stay consistent with P4 (P4 doesn't differentiate between the two).